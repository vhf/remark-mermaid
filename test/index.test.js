/* eslint-disable no-shadow */
const path = require('path');
const parse = require('remark-parse');
const stringify = require('remark-stringify');
const toVFile = require('to-vfile');
const unified = require('unified');
const mermaid = require('../src/');

const fixturesDir = path.join(__dirname, '/fixtures');
const runtimeDir = path.join(__dirname, '/runtime');
const remark = unified().use(parse).use(stringify).freeze();

// Utility function to add metdata to a vFile.
function addMetadata(vFile, destinationFilePath) {
  vFile.data = {
    destinationFilePath,
    destinationDir: path.dirname(destinationFilePath),
  };
}

jest.setTimeout(60000);

describe('remark-mermaid', () => {
  it('it ignores markdown that does not have mermaid references', () => {
    const srcFile = `${fixturesDir}/simple.md`;
    const destFile = `${runtimeDir}/simple.md`;
    const vfile = toVFile.readSync(srcFile);
    addMetadata(vfile, destFile);

    return remark().use(mermaid).process(vfile).then((vfile) => {
      expect(vfile.contents).not.toMatch(/!\[\]\(\.\/\w+\.svg/);
      expect(vfile.messages).toHaveLength(0);
    });
  });

  it('can handle code blocks', () => {
    const srcFile = `${fixturesDir}/code-block.md`;
    const destFile = `${runtimeDir}/code-block.md`;
    const vfile = toVFile.readSync(srcFile);
    addMetadata(vfile, destFile);

    return remark().use(mermaid).process(vfile).then((vfile) => {
      expect(vfile.contents).toMatch(/!\[\]\(\.\/\w+\.svg/);
      expect(vfile.messages[0].message).toBe('mermaid code block replaced with graph');
    });
  });

  it('can handle code blocks to svg string', () => {
    const srcFile = `${fixturesDir}/code-block.md`;
    const destFile = `${runtimeDir}/code-block.md`;
    const vfile = toVFile.readSync(srcFile);
    addMetadata(vfile, destFile);

    vfile.data.asString = true;

    return remark().use(mermaid).process(vfile).then((vfile) => {
      expect(vfile.contents).toMatch(/<svg id="[\s\S]*<\/svg>/);
      expect(vfile.messages[0].message).toBe('mermaid code block replaced with graph');
    });
  });

  it('can handle mermaid images', () => {
    const srcFile = `${fixturesDir}/image-mermaid.md`;
    const destFile = `${runtimeDir}/image-mermaid.md`;
    const vfile = toVFile.readSync(srcFile);
    addMetadata(vfile, destFile);

    return remark().use(mermaid).process(vfile).then((vfile) => {
      expect(vfile.contents).toMatch(/!\[Example\]\(\.\/\w+\.svg/);
      expect(vfile.messages[0].message).toBe('mermaid link replaced with link to graph');
    });
  });

  it('can handle mermaid links', () => {
    const srcFile = `${fixturesDir}/link-mermaid.md`;
    const destFile = `${runtimeDir}/link-mermaid.md`;
    const vfile = toVFile.readSync(srcFile);
    addMetadata(vfile, destFile);

    return remark().use(mermaid).process(vfile).then((file) => {
      expect(file.contents).toMatch(/\[Example\]\(\.\/\w+\.svg/);
      expect(vfile.messages[0].message).toBe('mermaid link replaced with link to graph');
    });
  });

  it('can handle a big mix of everything', () => {
    const srcFile = `${fixturesDir}/mix.md`;
    const destFile = `${runtimeDir}/mix.md`;
    const vfile = toVFile.readSync(srcFile);
    addMetadata(vfile, destFile);

    return remark().use(mermaid).process(vfile).then((vfile) => {
      expect(vfile.contents).toMatchSnapshot();
      const blocks = vfile.messages.filter(s => s.contains('with graph'));
      const links = vfile.messages.filter(s => s.contains('with link'));
      expect(blocks.length).toBe(4);
      expect(links.length).toBe(4);
    });
  });

  describe('simple mode', () => {
    it('can handle code blocks in simple mode', () => {
      const srcFile = `${fixturesDir}/code-block.md`;
      const destFile = `${runtimeDir}/code-block.md`;
      const vfile = toVFile.readSync(srcFile);
      addMetadata(vfile, destFile);

      return remark().use(mermaid, { simple: true }).process(vfile).then((vfile) => {
        expect(vfile.contents).toMatch(/class=\"mermaid\"/);
        expect(vfile.messages[0].message).toBe('mermaid code block replaced with div');
      });
    });

    it('can handle mermaid images in simple mode', () => {
      const srcFile = `${fixturesDir}/image-mermaid.md`;
      const destFile = `${runtimeDir}/image-mermaid.md`;
      const vfile = toVFile.readSync(srcFile);
      addMetadata(vfile, destFile);

      return remark().use(mermaid, { simple: true }).process(vfile).then((vfile) => {
        expect(vfile.contents).toMatch(/class=\"mermaid\"/);
        expect(vfile.messages[0].message).toBe('mermaid link replaced with div');
      });
    });

    it('can handle mermaid links in simple mode', () => {
      const srcFile = `${fixturesDir}/link-mermaid.md`;
      const destFile = `${runtimeDir}/link-mermaid.md`;
      const vfile = toVFile.readSync(srcFile);
      addMetadata(vfile, destFile);

      return remark().use(mermaid, { simple: true }).process(vfile).then((vfile) => {
        expect(vfile.contents).toMatch(/class=\"mermaid\"/);
        expect(vfile.messages[0].message).toBe('mermaid link replaced with div');
      });
    });
  });
});
