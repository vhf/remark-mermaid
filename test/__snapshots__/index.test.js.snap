// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`remark-mermaid can handle a big mix of everything 1`] = `
"# mermaid code block

![](./cc314291256b4fe90293ec53fdf97a00d652e5c1.svg \\"\`mermaid\` image\\")

# mermaid image

![Example](./bd582c6a57e1a09aa68906eb898c45f44f1cd6d6.svg \\"mermaid:\\")

# mermaid code block

![](./cc314291256b4fe90293ec53fdf97a00d652e5c1.svg \\"\`mermaid\` image\\")

![](./cc314291256b4fe90293ec53fdf97a00d652e5c1.svg \\"\`mermaid\` image\\")

\`\`\`md
graph LR
    Start --> Stop
\`\`\`

![](./cc314291256b4fe90293ec53fdf97a00d652e5c1.svg \\"\`mermaid\` image\\")

# mermaid link

-   [Example](./bd582c6a57e1a09aa68906eb898c45f44f1cd6d6.svg \\"mermaid:\\")
-   [Example](./bd582c6a57e1a09aa68906eb898c45f44f1cd6d6.svg \\"mermaid:\\")

# Lorem ipsum

Here is a paragraph with _some emphasis_ and some **strong text**, and a [link](http://google.com \\"title\\"). I ~~did~~ want to fix this.

\`\`\`bash
$ npm i
\`\`\`

-   ![Example](./bd582c6a57e1a09aa68906eb898c45f44f1cd6d6.svg \\"mermaid:\\")
-   ![Example](./bd582c6a57e1a09aa68906eb898c45f44f1cd6d6.svg \\"mermaid:\\")
"
`;

exports[`remark-mermaid can handle a big mix of everything 2`] = `
Array [
  [/Volumes/repositories/zds/zmarkdown/packages/remark-mermaid/test/fixtures/mix.md:37:3-37:36: mermaid link replaced with link to graph],
  [/Volumes/repositories/zds/zmarkdown/packages/remark-mermaid/test/fixtures/mix.md:36:3-36:36: mermaid link replaced with link to graph],
  [/Volumes/repositories/zds/zmarkdown/packages/remark-mermaid/test/fixtures/mix.md:14:1-17:4: mermaid code block replaced with graph],
  [/Volumes/repositories/zds/zmarkdown/packages/remark-mermaid/test/fixtures/mix.md:19:1-22:4: mermaid code block replaced with graph],
  [/Volumes/repositories/zds/zmarkdown/packages/remark-mermaid/test/fixtures/mix.md:29:1-32:4: mermaid code block replaced with graph],
  [/Volumes/repositories/zds/zmarkdown/packages/remark-mermaid/test/fixtures/mix.md:10:1-10:35: mermaid link replaced with link to graph],
  [/Volumes/repositories/zds/zmarkdown/packages/remark-mermaid/test/fixtures/mix.md:3:1-6:4: mermaid code block replaced with graph],
  [/Volumes/repositories/zds/zmarkdown/packages/remark-mermaid/test/fixtures/mix.md:47:3-47:37: mermaid link replaced with link to graph],
  [/Volumes/repositories/zds/zmarkdown/packages/remark-mermaid/test/fixtures/mix.md:48:3-48:37: mermaid link replaced with link to graph],
]
`;
